lib/ExecutionEngine/ExecutionEngine.cpp

```diff
@@ -234,6 +234,65 @@ void glow::runBatch(ExecutionEngine &EE, PlaceholderBindings &bindings,
   }
 }
 
+void glow::runBatchReportLoss(ExecutionEngine &EE,
+                              PlaceholderBindings &bindings, size_t iterations,
+                              size_t &sampleCounter,
+                              llvm::ArrayRef<Placeholder *> ph,
+                              llvm::ArrayRef<Tensor *> inputs,
+                              Placeholder *resultPH, std::vector<float> &loss,
+                              Placeholder *ceLossPH, bool verbose) {
+  // This is the size of one batch (the number of samples in the batch).
+  size_t batchSize = ph[0]->getType()->dims()[0];
+  loss.resize(iterations);
+
+  assert(!inputs.empty() && "No inputs");
+  assert(inputs.size() == ph.size() &&
+         "The number of inputs does not match the number of placeholders");
+
+  // For each iteration in the batch:
+  for (size_t j = 0; j < iterations; j++) {
+
+    // Update the input placeholders.
+    for (int i = 0, e = ph.size(); i < e; i++) {
+      assert(ph[i] && "Invalid value");
+      auto *backingTensor = bindings.get(ph[i]);
+      assert(backingTensor && "Can't find the backing tensor");
+
+      auto dim = inputs[i]->dims();
+      assert(backingTensor->dims().drop_front() == dim.drop_front() &&
+             "Invalid slice size");
+      // Extract the n'th slice, that must be a tensor.
+      size_t slc = sampleCounter % dim[0];
+      // Pick up one slice from the input tensors, and load it into the
+      // corresponding network Placeholder.
+      backingTensor->copyConsecutiveSlices(inputs[i], slc);
+    }
+
+    // Run the network.
+    EE.run(bindings);
+    sampleCounter += batchSize;
+
+    float ce_loss = 0;
```

Comment:
Instead of writing new method with duplicate code, you could modify the old method providing a callback method as an input and execute callback on every iteration:

```
void glow::runBatchWithCallback(ExecutionEngine &EE,
                              PlaceholderBindings &bindings, size_t iterations,
                              size_t &sampleCounter,
                              llvm::ArrayRef<Placeholder *> ph,
                              llvm::ArrayRef<Tensor *> inputs,
                              Placeholder *resultPH, 
                             std::function<void()> &&callback == nullptr) {
   loop:
      
     if (callback) {
       callback();
     }

}

```